### ISSUE TITLE
disklessset does not exit gracefully when failing

### ISSUE THREAD

**Issue by btravouillon** [bluebanquise/bluebanquise#296]
# Description

While reproducing an issue reported by a user with disklessset, I noticed the tool does not exit gracefully.
In the process, it creates and mounts a live image but does not umount it in case of failure in the process of image creation, leading to orphan files and eventually no space left in the filesystem.

# Info

OS: CentOS 8.1
Release: tag `1.3-rc1`

# Reproducer

```
[root@management1 bluebanquise]# df -h /root/diskless
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda1        10G  6.3G  4.7G  64% /

[root@management1 bluebanquise]# disklessset
BlueBanquise Diskless manager
 1 - List available kernels
 2 - Generate a new initramfs
 3 - Generate a new diskless image
 4 - Manage/list existing diskless images
 5 - Remove a diskless image
-->: 3

Starting new image creation phase.
Many questions will be asked, a recap will be provided before starting procedure.

Select image type:
 1 - nfs
 2 - livenet
-->: 2
[INFO] Loading kernels from /var/www/html/preboot_execution_environment/diskless/kernels/

Select kernel:
 1 - vmlinuz-4.18.0-147.3.1.el8_1.x86_64
-->: 1
Please enter image name ?
-->: test1
Please enter clear root password of the new image: test1
[INFO] Entering livenet dedicated part.
Please select livenet image generation profile:
 1 - Standard: core (~1.2Gb)
 2 - Small: openssh, dnf and NetworkManager (~248Mb)
 3 - Minimal: openssh only (~129Mb)
-->: 3
Please choose image size, considering /1000: 2M for 2Gb, 600K for 600Mb, etc:
-->: 2M
Do you want to create a new NFS image with the following parameters:
  Image name:           test1
  Kernel version:       vmlinuz-4.18.0-147.3.1.el8_1.x86_64
  Root password:        test1
  Image profile:        3
  Image size /1000:     2M
Confirm ? Enter yes or no: yes
[INFO] Cleaning and creating image folders.
[INFO] Generating new ipxe boot file.
[INFO] Creating empty image file, format and mount it.
[...]
[something goes wrong]
Traceback (most recent call last):
  File "/usr/bin/disklessset", line 332, in <module>
    with open('/mnt/etc/shadow') as ff:
FileNotFoundError: [Errno 2] No such file or directory: '/mnt/etc/shadow'
```
Available space decreased by 2 GB. /mnt is still mounted.
```
[root@management1 bluebanquise]# df -h /root/diskless
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda1        10G  7.3G  2.7G  74% /
[root@management1 bluebanquise]# df -h /mnt
Filesystem      Size  Used Avail Use% Mounted on
/dev/loop0      2.0G   47M  2.0G   3% /mnt
```
Second run with the same parameters, same crash of the application.

```
[root@management1 bluebanquise]# disklessset
[...]
FileNotFoundError: [Errno 2] No such file or directory: '/mnt/etc/shadow'
[root@management1 bluebanquise]# df -h /root/diskless
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda1        10G  9.3G  710M  94% /
[root@management1 bluebanquise]# df -h /mnt
Filesystem      Size  Used Avail Use% Mounted on
/dev/loop1      2.0G   47M  2.0G   3% /mnt
```
Now we see /dev/loop1 mounted in /mnt, but /dev/loop0 is still mounted.
```
[root@management1 bluebanquise]# df /dev/loop0
df: cannot access '/dev/loop0': over-mounted by another device
[root@management1 bluebanquise]# losetup /dev/loop0
/dev/loop0: [64769]:4936449 (/root/diskless/workdir/LiveOS/rootfs.img (deleted))
[root@management1 bluebanquise]# losetup /dev/loop1
/dev/loop1: [64769]:19917969 (/root/diskless/workdir/LiveOS/rootfs.img)
```

Third run with the same parameters, same crash of the application. Note the size of the image does not match the requirement, but the script continues:
```
[root@management1 bluebanquise]# disklessset
[...]
[INFO] Creating empty image file, format and mount it.
dd: error writing '/root/diskless/workdir//LiveOS/rootfs.img': No space left on device
726153+0 records in
726152+0 records out
743579648 bytes (744 MB, 709 MiB) copied, 1.4847 s, 501 MB/s
[...]
Traceback (most recent call last):
  File "/usr/bin/disklessset", line 332, in <module>
    with open('/mnt/etc/shadow') as ff:
FileNotFoundError: [Errno 2] No such file or directory: '/mnt/etc/shadow'
```

The file system is full:
```
[root@management1 bluebanquise]# df -h /root/diskless
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda1        10G   10G   20K 100% /
[root@management1 bluebanquise]# df -h /mnt
Filesystem      Size  Used Avail Use% Mounted on
/dev/loop2      704M   38M  666M   6% /mnt
[root@management1 bluebanquise]# losetup /dev/loop0; losetup /dev/loop1; losetup /dev/loop2
/dev/loop0: [64769]:4936449 (/root/diskless/workdir/LiveOS/rootfs.img (deleted))
/dev/loop1: [64769]:19917969 (/root/diskless/workdir/LiveOS/rootfs.img (deleted))
/dev/loop2: [64769]:4936470 (/root/diskless/workdir/LiveOS/rootfs.img)
```

# Workaround
```
[root@management1 bluebanquise]# umount /dev/loop2 /dev/loop1 /dev/loop0
[root@management1 bluebanquise]# df -h /root/diskless
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda1        10G  6.0G  4.0G  60% /
[root@management1 bluebanquise]# ls -lh /root/diskless/workdir/LiveOS/rootfs.img
-rw-r--r--. 1 root root 710M Jul 29 01:22 /root/diskless/workdir/LiveOS/rootfs.img
[root@management1 bluebanquise]# rm -f /root/diskless/workdir/LiveOS/rootfs.img
```

**Comment by btravouillon** [comment]
It does not clean `/var/www/html/preboot_execution_environment/diskless/images/` neither, which lead to application crash when listing available images:

```
# disklessset 
BlueBanquise Diskless manager
 1 - List available kernels
 2 - Generate a new initramfs
 3 - Generate a new diskless image
 4 - Manage/list existing diskless images
 5 - Remove a diskless image
-->: 4
[INFO] Entering images management part.
 1 - List available images
 2 - Manage kernel of an image
 3 - Create a golden from a staging NFS image
 4 - Manage hosts of an NFS image
-->: 1

  Image name: test1
Traceback (most recent call last):
  File "/usr/bin/disklessset", line 367, in <module>
    with open('/var/www/html/preboot_execution_environment/diskless/images/'+str(i)+'/image_data.yml', 'r') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/var/www/html/preboot_execution_environment/diskless/images/test1/image_data.yml'

# ls /var/www/html/preboot_execution_environment/diskless/images/test1/
boot.ipxe
```

**Comment by oxedions** [comment]
Hello
Very late.
A new version of the script, fully revamped, is under development. I am waiting for this version to be released to include such "auto cleaning" parts.

**Comment by btravouillon** [comment]
With current master branch (c0fe129cf6cccce4802324dc1d7f33bbe3e7f859), a second run fails earlier with:

```
[WARNING] The directory /var/tmp/diskless/workdir/fail already exists. Cleaning.
Traceback (most recent call last):
  File "/usr/bin/disklessset", line 368, in <module>
    os.makedirs(image_working_directory)
  File "/usr/lib64/python3.6/os.py", line 220, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: '/var/tmp/diskless/workdir/fail'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/disklessset", line 371, in <module>
    shutil.rmtree(image_working_directory)
  File "/usr/lib64/python3.6/shutil.py", line 486, in rmtree
    _rmtree_safe_fd(fd, path, onerror)
  File "/usr/lib64/python3.6/shutil.py", line 428, in _rmtree_safe_fd
    onerror(os.rmdir, fullname, sys.exc_info())
  File "/usr/lib64/python3.6/shutil.py", line 426, in _rmtree_safe_fd
    os.rmdir(name, dir_fd=topfd)
OSError: [Errno 16] Device or resource busy: 'mnt'
```

This prevents to create additional rootfs images and loop devices, which solves the ENOSPACE condition.

The mount point is now clearly identified:
```
# df -h | grep diskless
/dev/loop0               2.0G   47M  2.0G   3% /var/tmp/diskless/workdir/fail/mnt
```

The workaround is:
```
# umount /dev/loop0
```

I would like to keep this ticket opened for possible fixes in 1.3.1. (at least exit gracefully instead of printing a stack trace. Let's target auto cleaning for 1.4+)

**Comment by oxedions** [comment]
Ok for me 🙂

**Comment by btravouillon** [comment]
Also, don't just hit enter here:
```
Select kernel:
 1 - vmlinuz-4.18.0-193.1.2.el8_2.x86_64
-->:
Traceback (most recent call last):
  File "/usr/bin/disklessset", line 254, in <module>
    selected_kernel = select_from_list(kernel_list, 'kernel', -1)
  File "/usr/bin/disklessset", line 65, in select_from_list
    selected_item = str(int(input('-->: ').lower().strip())+index_modifier)
ValueError: invalid literal for int() with base 10: '' 
```

**Comment by btravouillon** [comment]
Bug report was fixed in 1.3.0.
The 3 bugs reported in comments are now fixed in `master` and `stable-1.3`.
